core/sqlx_parser.ts

```diff
@@ -0,0 +1,257 @@
+  CAPTURE_EVERYTHING_ELSE: LEXER_STATE_NAMES.INNER_SQL_BLOCK + "_captureEverythingElse"
+};
+
+const lexer = moo.states(buildSqlxLexer());
+
+export interface ISqlxParseResults {
+  config: string;
+  js: string;
+  sql: string[];
+  incremental: string;
+  preOperations: string[];
+  postOperations: string[];
+}
+
+export function parseSqlx(code: string): ISqlxParseResults {
+  const results = {
+    config: "",
+    js: "",
+    sql: [""],
+    incremental: "",
+    preOperations: [""],
+    postOperations: [""]
+  };
+  const parseState = new SqlxParseState();
+  lexer.reset(code);
+  for (const token of lexer) {
+    const state = parseState.computeState(token);
+    if (
+      token.type === SQL_LEXER_TOKEN_NAMES.STATEMENT_SEPERATOR ||
+      token.type === INNER_SQL_BLOCK_LEXER_TOKEN_NAMES.STATEMENT_SEPERATOR
+    ) {
+      if (state === "incremental") {
+        throw new Error(
+          "Incremental code blocks may not contain SQL statement separators ('---')."
+        );
+      }
+      (results[state] as string[]).push("");
+    }
+    if (Array.isArray(results[state])) {
+      (results[state] as string[])[results[state].length - 1] += token.value;
+    } else {
+      results[state] += token.value;
+    }
+  }
+  if (results.js) {
+    // If the user provided any JS, cut off the last closing brace.
+    // We have to do this because we intentionally cut off the starting brace during lexing.
+    // We can't keep them because the user's JS must not be run inside a scoped block.
+    results.js = results.js.substring(0, results.js.length - 1);
+  }
+  return results;
+}
+
+const tokenTypeStateMapping = new Map<string, keyof ISqlxParseResults>();
+tokenTypeStateMapping.set(SQL_LEXER_TOKEN_NAMES.START_CONFIG, "config");
+tokenTypeStateMapping.set(SQL_LEXER_TOKEN_NAMES.START_JS, "js");
+tokenTypeStateMapping.set(SQL_LEXER_TOKEN_NAMES.START_INCREMENTAL, "incremental");
+tokenTypeStateMapping.set(SQL_LEXER_TOKEN_NAMES.START_PRE_OPERATIONS, "preOperations");
+tokenTypeStateMapping.set(SQL_LEXER_TOKEN_NAMES.START_POST_OPERATIONS, "postOperations");
+
+class SqlxParseState {
+  private currentState: keyof ISqlxParseResults = "sql";
+
+  public computeState(token: moo.Token): keyof ISqlxParseResults {
+    if (!token.type.startsWith(LEXER_STATE_NAMES.SQL)) {
+      return this.currentState;
+    }
+    this.currentState = "sql";
+    if (tokenTypeStateMapping.has(token.type)) {
+      this.currentState = tokenTypeStateMapping.get(token.type);
+    }
+    return this.currentState;
+  }
+}
+
+function buildSqlxLexer() {
+  const sqlLexer = {};
+  sqlLexer[SQL_LEXER_TOKEN_NAMES.START_CONFIG] = {
+    match: "config {",
+    push: LEXER_STATE_NAMES.JS_BLOCK,
+    value: () => "{"
+  };
+  sqlLexer[SQL_LEXER_TOKEN_NAMES.START_JS] = {
+    match: "js {",
+    push: LEXER_STATE_NAMES.JS_BLOCK,
+    value: () => ""
+  };
+  sqlLexer[SQL_LEXER_TOKEN_NAMES.START_INCREMENTAL] = {
+    match: "if_incremental {",
```

Comment:
Ok so I think I got ahead of myself here. This block should probably be called `incremental_where` or just `where`.

Longer term, the idea here was that you would write something like:

```
select user_id, sum(1) from test
if_incremental {
  where ts > (select max(ts) from ${self()}
}
group by 1
```

The idea being you can control exactly where the conditional lines go. We need a completely new proto structure to do something like this for blocks of SQL. As it stands we currently just wrap the query in a select statement and apply the where clause, and I don't think the if_incremental name makes that very clear (plus we should save that name for when we do the new thing!).